Add userCourse type

jessehartloff opened the pull request:
## Proposed changes
--------------------
Added the UserCourse type to the shared types.

## Types of changes
--------------------
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist
--------------------
- [x] My changeset covers only what is described above (no extraneous changes)
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged downstream

jessehartloff committed:
Add UserCourse type

jessehartloff committed:
Add dropped to UserCourse type

jessehartloff committed:
Change userId from type number to type User

jessehartloff committed:
Revert userId back to type number

NicholasMy commented on src/index.ts:
```diff
@@ -2,6 +2,7 @@
 
 export * from './types/auth.types'
 export * from './types/user.types'
+export * from './types/user-course.type'
```
We probably want to stick with the `.types` extension for consistency.

MikeDrewitt commented on src/types/user-course.type.ts:
```diff
@@ -0,0 +1,10 @@
+  userId: number
+  courseId: number
+  dropped?: boolean
+  level?: string
```
If this can be only certain strings then this type should only be 

`'instructor' | 'ta' | 'student'`

jessehartloff replied:
You can do that?? TpyeScript ftw!

jessehartloff committed:
Updates based on PR comments
-Add 's' to "type" in user-course.type
-Change the type of level from string to a list of valid string literals

jessehartloff committed:
Update levels to be a type instead of a list of literals

jessehartloff committed:
export userCourseLevels

jessehartloff commented:
Discussion in related PR https://github.com/UBAutograding/devU-api/pull/18#discussion_r662389485

MikeDrewitt commented on src/types/user-course.types.ts:
```diff
@@ -0,0 +1,26 @@
+
+// I was unable to find a way to remove the duplication here. I want to create a type based on the
+// userCourseRoles.value to avoid having each value in 2 places. If anyone knows how to do this, please share!
+export const userCourseLevels = ['student', 'ta', 'instructor'] as const
```
```suggestion
export const userCourseLevels = userCourseRoles.map(u => u.value) as const
```

jessehartloff replied:
I get this error when I try that
`TS1355: A 'const' assertions can only be applied to references to enum members, or string, number, boolean, array, or object literals.`

jessehartloff committed:
dropped and level are no longer optional

jessehartloff committed:
typo

jessehartloff committed:
cleanup

jessehartloff committed:
cleanup

MikeDrewitt committed:
lecture section is nullable not optional

MikeDrewitt committed:
Change import

MikeDrewitt committed:
Remove lecture section
